give this project some more care

Techatrix opened the pull request:
No clue what I should name this PR so please forgive me.
The individual commits should give a good overview of everything I have changed but here are some of the highlights:

- add doc comments to the `KnownFolder` enum
- add an installation guide in the README
- port `xdg-user-dir-lookup.c` from xdg-user-dirs to Zig which takes care of properly parsing the `user-dirs.dirs` file.
- add tests for linux, macos and wasm

I have abstracted away all OS interactions to test different environment variables and `user-dirs.dirs` files. It can also test different comptime configs (`KnownFolderConfig`).

The public API remains unchanged.

Techatrix commented:
Anyone out here who would like to review this?

ikskuh commented on build.zig:
```diff
@@ -1,7 +1,44 @@
 const std = @import("std");
+const builtin = @import("builtin");
+
+const minimum_zig_version = std.SemanticVersion.parse("0.12.0") catch unreachable;
```
You can also mark that visible in the build.zig.zon

Techatrix replied:
e8b288e57de5c67b49d9d665a2261f6219127dd0

ikskuh commented on known-folders.zig:
```diff
@@ -219,140 +569,598 @@ const WindowsFolderSpec = union(enum) {
+        .fonts => "Library/Fonts",
+        .app_menu => "Applications",
+        .cache => "Library/Caches",
+        .roaming_configuration => "Library/Preferences",
+        .local_configuration => "Library/Application Support",
+        .global_configuration => "/Library/Preferences", // absolute path
```
Is the missing leading `/` wanted here?

Techatrix replied:
Which missing `/` are you referring to?
I didn't touch any of the known folder specifications. This is identical to the old version:
https://github.com/ziglibs/known-folders/blob/47076c6b11214a218e9244471d8762310820911a/known-folders.zig#L327-L329

ikskuh replied:
Ah yes. Sorry, i totally had the oversight of `.absolute` here.

ikskuh commented:
Looks okay overall, but pretty large to review. I assume it works on your machine?

Techatrix commented:
> Looks okay overall, but pretty large to review. I assume it works on your machine?

Yes, I have tested this locally on Linux. The windows specific code is mostly unchanged. The tests that I have added run for linux, macos and wasi. See 515e1f52ba4b49719f5ca3f4383a25983f417cc9. They even contain some macos specific tests that I never ran locally.

Should I split up this PR to make it easier to review? The whole diff looks very messy but I tried to make the individual commit readable as possible. With limited success. Please let me know if there is anything I can do to make this easier to review.

ikskuh reviewed (approved):
I'm pretty happy with it as-is for :)

ikskuh commented:
@Techatrix shall we merge?
